api_app/services/aad_authentication.py

```diff
@@ -278,21 +279,31 @@ def _get_users_inc_groups_from_response(self, users_graph_data, roles_graph_data
                 user_name = user_data["body"]["displayName"]
 
                 if "users" in user_data["body"]["@odata.context"]:
-                    user_email = user_data["body"]["mail"]
+                    user_email = user_data["body"]["userPrincipalName"]
                     # if user with id does not already exist in users
+                    user_roles = self._get_roles_for_principal(user_id, roles_graph_data, app_id_to_role_name)
+
                     if not any(user.id == user_id for user in users):
-                        users.append(User(id=user_id, name=user_name, email=user_email, roles=self._get_roles_for_principal(user_id, roles_graph_data, app_id_to_role_name)))
+                        users.append(User(id=user_id, name=user_name, email=user_email, roles=user_roles))
+                    else:
+                        user = next((user for user in users if user.id == user_id), None)
+                        user.roles = list(set(user.roles + user_roles))
```

Comment:
Whats happening to this user object, does it need appending somewhere?

Reply:
It is there to ensure that if we have a combination of app roles and groups the list reflects that.